hutch_python/log_setup.py

```diff
@@ -217,36 +226,227 @@ class ObjectFilter(logging.Filter):
+        If a single ophyd object logs over ``noisy_threshold_60s`` log messages
+        in 60 seconds, consider it a noisy logger and silence it.
+
+    whitelist : list of str, optional
+        Logger or object names that are not subject to the thresholds above.
+
+    blacklist : list of str, optional
+        Logger or object names that should always be filtered out.
+
+    Attributes
+    ----------
+    blacklist : list of str
+        Logger names that should always be filtered out.
+
+    whitelist : list of str
+        List of noisy loggers or object names that are exempt from the noise
+        thresholds.
+
+    noisy_loggers: set of str
+        Loggers marked as noisy and to be filtered out, unless in the
+        whitelist.
+    """
+    _objects: frozenset[ophyd.ophydobj.OphydObject]
+    _timer: threading.Thread
+    _running: bool
+    level: str
+    allow_other_messages: bool
+    whitelist_all_level: str
+    name_to_log_count_1s: dict[str, int]
+    name_to_log_count_10s: dict[str, int]
+    name_to_log_count_60s: dict[str, int]
+    noisy_threshold_1s: int
+    noisy_threshold_10s: int
+    noisy_threshold_60s: int
+    whitelist: list[str]
+    noisy_loggers: set[str]
+    blacklist: list[str]
+
+    def __init__(
+        self,
+        *objects: ophyd.ophydobj.OphydObject,
+        level: str = "WARNING",
+        whitelist_all_level: str = "WARNING",
+        noisy_threshold_1s: int = 20,
+        noisy_threshold_10s: int = 50,
+        noisy_threshold_60s: int = 100,
+        whitelist: Optional[list[str]] = None,
+        blacklist: Optional[list[str]] = None,
+        allow_other_messages: bool = True
+    ):
         self._objects = frozenset(objects)
         self.allow_other_messages = bool(allow_other_messages)
         self.whitelist_all_level = whitelist_all_level
         self.level = level
+        self.name_to_log_count_1s = collections.defaultdict(int)
+        self.name_to_log_count_10s = collections.defaultdict(int)
+        self.name_to_log_count_60s = collections.defaultdict(int)
+        self.noisy_threshold_1s = int(noisy_threshold_1s)
+        self.noisy_threshold_10s = int(noisy_threshold_10s)
+        self.noisy_threshold_60s = int(noisy_threshold_10s)
+        self.whitelist = list(whitelist or [])
+        self.blacklist = list(blacklist or [])
+        self.noisy_loggers = set()
+
+        self._running = True
+        self._timer_index = 0
+
+        self._timer = threading.Thread(target=self._count_update_thread)
+        self._timer.daemon = True
+        self._timer.start()
+
+    def _count_update_thread(self) -> None:
+        """Thread for checking the per-logger statistics."""
+        while self._running:
+            self._count_update()
+            time.sleep(1.0)
+
+    def _count_update(self) -> None:
+        """Log count update - every second."""
+        noisy_loggers = set(
+            name
+            for name, count in tuple(self.name_to_log_count_1s.items())
```

Comment:
is the `tuple` here just to eliminate cross-thread dictionary edits while iterating?

Reply:
Yes, loggers recording messages in other threads could mutate the dictionary. Alternative would be a `threading.Lock` wrapping around these, but it's a bit costlier.